routing/state.go

```diff
@@ -0,0 +1,157 @@
+package routing
+
+import (
+	"errors"
+	"github.com/gogo/protobuf/proto"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	pb "github.com/libp2p/go-libp2p-core/routing/pb"
+	ma "github.com/multiformats/go-multiaddr"
+	"time"
+)
+
+// The domain string used for routing state records contained in a SignedEnvelope.
+const StateEnvelopeDomain = "libp2p-routing-state"
+
+// The type hint used to identify routing state records in a SignedEnvelope.
+// TODO: register multicodec
+var StateEnvelopePayloadType = []byte("/libp2p/routing-state-record")
+
+type SignedRoutingState struct {
+	// PeerID is the ID of the peer this record pertains to.
+	PeerID peer.ID
+
+	// Seq is an increment-only sequence counter used to order RoutingState records in time.
+	Seq uint64
```

Comment:
Once again, why is the seq _inside_ the payload? I don't regard this as part of the payload itself, and it's very likely that all records need version management, so it makes sense to move it to the outer layer.

If it's for signing purposes, we can concatenate the sequence number to the payload plaintext prior to signing.

Reply:
I think @Stebalien's objection was that it makes the envelope less generic. I think I'd prefer to have it on the outer envelope also, since it would let you quickly reject old records without having to deserialize the payload.